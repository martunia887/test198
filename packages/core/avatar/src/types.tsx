import { AnalyticsEventInterface } from '@atlaskit/analytics-next';
import { ThemeProp } from '@atlaskit/theme';
import { ReactNode, ComponentType } from 'react';
import { ThemeProps, ThemeTokens } from './theme';

export enum AppearanceType {
  CIRCLE = 'circle',
  SQUARE = 'square',
}

export type PresenceType =
  | ('online' | 'busy' | 'focus' | 'offline')
  | ReactNode;

export type Size =
  | 'xsmall'
  | 'small'
  | 'medium'
  | 'large'
  | 'xlarge'
  | 'xxlarge';

// NOTE: this won't work as an interface
// NOTE: this only works if key is optional
export type SizeType = { [key in Size]?: number };

// NOTE: sizes xsmall & xxlarge DO NOT support
// - groups
// - presence
// - status

export type SupportedSizeWithAnIcon = Pick<
  SizeType,
  Exclude<keyof SizeType, 'xsmall' | 'xxlarge'>
>;

export type StatusType = ('approved' | 'declined' | 'locked') | ReactNode;
export type StyledComponentType = 'custom' | 'button' | 'link' | 'span';

export type AvatarClickType = (
  event?: { event?: KeyboardEvent | MouseEvent; item: Object },
  analyticsEvent?: AnalyticsEventInterface,
) => void;

export interface AvatarPropTypesBase {
  /** Indicates the shape of the avatar. Most avatars are circular, but square avatars
   can be used for 'container' objects. */
  appearance: AppearanceType;
  /** Defines the size of the avatar */
  size: SizeType;
  /** Display a tooltip on hover */
  enableTooltip: boolean;
  /** Used to override the default border color of the presence indicator.
   Accepts any color argument that the border-color CSS property accepts. */
  borderColor?: string | (() => any);
  /** A custom component to use instead of the default span.
   * A `className` prop is passed to the component which has classNames for all the default styles for the avatar.
   * */
  component?: ComponentType<any>;

  /** Provides a url for avatars being used as a link. */
  href?: string;
  /** Change the style to indicate the avatar is active. */
  isActive?: boolean;
  /** Change the style to indicate the avatar is disabled. */
  isDisabled?: boolean;
  /** Change the style to indicate the avatar is focused. */
  isFocus?: boolean;
  /** Change the style to indicate the avatar is hovered. */
  isHover?: boolean;
  /** Change the style to indicate the avatar is selected. */
  isSelected?: boolean;
  /** Name will be displayed in a tooltip, also used by screen readers as fallback
   content if the image fails to load. */
  name?: string;
  /** Indicates a user's online status by showing a small icon on the avatar.
  Refer to presence values on the Presence component.
  Alternatively accepts any React element. For best results, it is recommended to
  use square content with height and width of 100%. */
  presence?: PresenceType;

  /** A url to load an image from (this can also be a base64 encoded image). */
  src?: string;
  /** Indicates contextual information by showing a small icon on the avatar.
   Refer to status values on the Status component. */
  status?: StatusType;
  /** The index of where this avatar is in the group `stack`. */
  stackIndex?: number;
  /** Assign specific tabIndex order to the underlying node. */
  tabIndex?: number;
  /** Pass target down to the anchor, if href is provided. */
  target?: '_blank' | '_self' | '_top' | '_parent';
  /** You should not be accessing this prop under any circumstances. It is
   provided by @atlaskit/analytics-next. */
  createAnalyticsEvent?: any;

  /** The theme that should be applied to the avatar. */
  theme?: ThemeProp<ThemeTokens, ThemeProps>;
}

export interface AvatarPropTypes extends AvatarPropTypesBase {
  /** Handler to be called on click. */
  onClick?: AvatarClickType;
}

export interface SkeletonProps {
  /* Incidcates the shape of the skeleton */
  appearance: AppearanceType;
  /* Sets the color of the skeleton. By default it will inherit the current text color. */
  color?: string;
  /* Defines the size of the skeleton */
  size: Size;
  /* Determines the opacity of the skeleton */
  weight: 'normal' | 'strong';
}
