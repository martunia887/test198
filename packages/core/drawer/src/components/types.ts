import { ComponentType, SyntheticEvent, ReactChild } from 'react';
import {
  WithAnalyticsEventProps,
  UIAnalyticsEvent,
} from '@atlaskit/analytics-next';

export type Widths = {
  full: string;
  extended: string;
  narrow: number;
  medium: number;
  wide: number;
};

export type DrawerWidth = keyof Widths;

export type IconProps = { size: string };

export type BaseProps = {
  /** The content of the drawer */
  children: ReactChild;
  /** Icon to be rendered in your drawer as a component, if available */
  icon?: ComponentType<IconProps>;
  /** Available drawer sizes */
  width: DrawerWidth;
};

export type DrawerPrimitiveProps = BaseProps & {
  in: boolean;
  onClose?: (event: SyntheticEvent) => void;
  onCloseComplete?: (node: HTMLElement) => void;
  shouldUnmountOnExit?: boolean;
};

export type DrawerProps = BaseProps &
  WithAnalyticsEventProps & {
    /**
      Callback function to be called when the drawer will be closed.
    */
    onClose?: (
      event: SyntheticEvent,
      analyticsEvent?: UIAnalyticsEvent,
    ) => void;
    /** A callback function that will be called when the drawer has finished its close transition. */
    onCloseComplete?: (node: HTMLElement) => void;
    /**
      Callback function that will be called when the drawer is displayed and `keydown` event is triggered.
    */
    onKeyDown?: (event: SyntheticEvent) => void;
    /** Controls if the drawer is open or close */
    isOpen: boolean;
    /** Boolean that controls if drawer should be retained/discarded */
    shouldUnmountOnExit?: boolean;
  };

/**
  Type of keyboard event that triggers which key will should close the drawer.
*/
export type CloseTrigger = 'backButton' | 'blanket' | 'escKey';
