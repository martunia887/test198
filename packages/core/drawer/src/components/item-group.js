// @flow

import React, { type Node } from 'react';
import { ItemGroup } from '@atlaskit/item';
import { fontSizeSmall, gridSize as gridSizeFn } from '@atlaskit/theme';

type Props = {
  /** React Elements to be displayed within the group. This should generally be
   a collection of NavigationItems. */
  children?: Node,
  /** Set whether the text should be compacted. */
  isCompact?: boolean,
  /** Text to appear as heading above group. Will be auto-capitalised. */
  title?: string,
};

const groupTitleFontSize = fontSizeSmall();
const gridSize = gridSizeFn();

const SkeletonGroupTitle = ({ ...props }) => (
  <div
    css={{
      fontSize: `${groupTitleFontSize}px`,
      lineHeight: `${(gridSize * 2) / groupTitleFontSize}`,
      fontWeight: 600,
      marginTop: `${gridSize * 1.5}px`,
    }}
    {...props}
  />
);
const ItemChildrenWrapper = ({ ...props }) => (
  <div css={{ marginLeft: `${gridSize}px` }} {...props} />
);
const ItemGroupWrapper = ({ ...props }) => (
  <div css={{ paddingRight: `${gridSize * 4}px` }} {...props} />
);

export const DrawerItemGroup = ({ title, isCompact, children }: Props) => {
  const wrappedTitle = title ? (
    <SkeletonGroupTitle>{title}</SkeletonGroupTitle>
  ) : null;

  return (
    <ItemGroupWrapper>
      <ItemGroup title={wrappedTitle} isCompact={isCompact}>
        <ItemChildrenWrapper>{children}</ItemChildrenWrapper>
      </ItemGroup>
    </ItemGroupWrapper>
  );
};
