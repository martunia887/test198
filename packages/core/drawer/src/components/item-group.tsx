import React, { FC } from 'react';
import { ItemGroup } from '@atlaskit/item';
import { gridSize as gridSizeFn } from '@atlaskit/theme';
import { fontSizeSmall } from '@atlaskit/theme/constants';

type Props = {
  /** Set whether the text should be compacted. */
  isCompact?: boolean;
  /** Text to appear as heading above group. Will be auto-capitalised. */
  title?: string;
};

const groupTitleFontSize = fontSizeSmall();
const gridSize = gridSizeFn();

const SkeletonGroupTitle = ({ ...props }) => (
  <div
    css={{
      fontSize: `${groupTitleFontSize}px`,
      lineHeight: `${(gridSize * 2) / groupTitleFontSize}`,
      fontWeight: 600,
      marginTop: `${gridSize * 1.5}px`,
    }}
    {...props}
  />
);
const ItemChildrenWrapper = ({ ...props }) => (
  <div css={{ marginLeft: `${gridSize}px` }} {...props} />
);
const ItemGroupWrapper = ({ ...props }) => (
  <div css={{ paddingRight: `${gridSize * 4}px` }} {...props} />
);

export const DrawerItemGroup: FC<Props> = ({ title, isCompact, children }) => {
  const wrappedTitle = title ? (
    <SkeletonGroupTitle>{title}</SkeletonGroupTitle>
  ) : null;

  return (
    <ItemGroupWrapper>
      <ItemGroup title={wrappedTitle} isCompact={isCompact}>
        <ItemChildrenWrapper>{children}</ItemChildrenWrapper>
      </ItemGroup>
    </ItemGroupWrapper>
  );
};
