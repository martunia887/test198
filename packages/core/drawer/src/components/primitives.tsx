/** @jsx jsx */

import { Component, FC, MouseEvent } from 'react';
import { layers, gridSize } from '@atlaskit/theme/constants';
import { N0, N500, N30A, B50 } from '@atlaskit/theme/colors';
import ArrowLeft from '@atlaskit/icon/glyph/arrow-left';
import { jsx } from '@emotion/core';

import { DrawerPrimitiveProps, DrawerWidth, Widths } from './types';
import { SlideIn, ExitingPersistence } from '@atlaskit/motion';

// Misc.
// ------------------------------

const widths: Widths = {
  full: '100vw',
  extended: '95vw',
  narrow: 45 * gridSize(),
  medium: 60 * gridSize(),
  wide: 75 * gridSize(),
};

// Wrapper
// ------------------------------

const Wrapper = ({
  width = 'narrow',
  shouldUnmountOnExit,
  ...props
}: {
  children?: React.ReactNode;
  shouldUnmountOnExit?: boolean;
  width: DrawerWidth;
}) => {
  return (
    <div
      css={{
        backgroundColor: N0,
        display: 'flex',
        height: '100vh',
        left: 0,
        overflow: 'hidden',
        position: 'fixed',
        top: 0,
        width: widths[width],
        zIndex: layers.blanket() + 1,
      }}
      {...props}
    />
  );
};

// Content
// ------------------------------

const Content: FC = props => (
  <div
    css={{ flex: 1, marginTop: 3 * gridSize(), overflow: 'auto' }}
    {...props}
  />
);

// Sidebar / Icons etc.
// ------------------------------

const Sidebar: FC = props => {
  return (
    <div
      css={{
        alignItems: 'center',
        boxSizing: 'border-box',
        color: N500,
        display: 'flex',
        flexShrink: 0,
        flexDirection: 'column',
        height: '100vh',
        paddingBottom: 2 * gridSize(),
        paddingTop: 3 * gridSize(),
        width: 8 * gridSize(),
      }}
      {...props}
    />
  );
};

interface IconWrapperProps {
  onClick?: (event: MouseEvent<HTMLElement>) => void;
}
const IconWrapper: FC<IconWrapperProps> = props => (
  <button
    type="button"
    css={{
      alignItems: 'center',
      background: 0,
      border: 0,
      borderRadius: '50%',
      color: 'inherit',
      cursor: props.onClick ? 'pointer' : undefined,
      display: 'flex',
      fontSize: 'inherit',
      height: 5 * gridSize(),
      justifyContent: 'center',
      lineHeight: 1,
      marginBottom: 2 * gridSize(),
      padding: 0,
      width: 5 * gridSize(),

      '&:hover': {
        backgroundColor: props.onClick ? N30A : undefined,
      },
      '&:active': {
        backgroundColor: props.onClick ? B50 : undefined,
        outline: 0,
      },
    }}
    {...props}
  />
);

export default class DrawerPrimitive extends Component<DrawerPrimitiveProps> {
  render() {
    const { children, icon: Icon, onClose, ...props } = this.props;

    const onCloseComplete =
      this.props.onCloseComplete !== undefined
        ? this.props.onCloseComplete
        : () => {};

    const onOpenComplete =
      this.props.onOpenComplete !== undefined
        ? this.props.onOpenComplete
        : () => {};

    return (
      <ExitingPersistence>
        {props.in && (
          <SlideIn
            onFinish={direction => {
              direction === 'entering' ? onOpenComplete() : onCloseComplete();
            }}
            from="left"
          >
            {pr => {
              return (
                <Wrapper
                  width={props.width !== undefined ? props.width : 'medium'}
                  {...pr}
                >
                  <Sidebar>
                    <IconWrapper
                      onClick={onClose}
                      data-test-selector="MotionDrawerPrimitiveSidebarCloseButton"
                    >
                      {Icon ? (
                        <Icon size="large" />
                      ) : (
                        <ArrowLeft label="Close drawer" />
                      )}
                    </IconWrapper>
                  </Sidebar>
                  <Content>{children}</Content>
                </Wrapper>
              );
            }}
          </SlideIn>
        )}
      </ExitingPersistence>
    );
  }
}
