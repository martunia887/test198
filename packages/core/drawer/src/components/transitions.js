// @flow

import React, { Component, type ComponentType } from 'react';
import { Transition } from 'react-transition-group';
import { layers } from '@atlaskit/theme';

import { transitionDurationMs, transitionTimingFunction } from '../constants';

// Transitions
// ------------------------------

type Styles = { [string]: string | number | null };

type TransitionProps = {
  children?: Node,
  component?: ComponentType<*> | string,
  onExited?: (node: HTMLElement) => void,
  openFromRight: boolean,
  shouldUnmountOnExit?: boolean,
  in: boolean,
};

type HandlerProps = {
  defaultStyles: Styles,
  transitionProps: {
    appear: boolean,
    mountOnEnter: boolean,
    unmountOnExit: boolean,
  },
  transitionStyles: {
    entering?: Styles,
    entered?: Styles,
    exiting?: Styles,
    exited?: Styles,
  },
};

const defaultTransitionProps = {
  appear: true,
  mountOnEnter: true,
  unmountOnExit: true,
};
class TransitionHandler extends Component<TransitionProps & HandlerProps> {
  static defaultProps = {
    component: 'div',
    transitionProps: defaultTransitionProps,
  };

  render() {
    const {
      component: Tag = 'div',
      in: inProp,
      onExited,
      defaultStyles,
      transitionStyles,
      transitionProps,
      ...props
    } = this.props;
    const timeout = { enter: 0, exit: transitionDurationMs };

    return (
      <Transition
        in={inProp}
        onExited={onExited}
        timeout={timeout}
        {...transitionProps}
      >
        {state => {
          const style = {
            ...defaultStyles,
            ...transitionStyles[state],
          };

          return <Tag style={style} {...props} />;
        }}
      </Transition>
    );
  }
}

export const Fade = ({ ...props }: TransitionProps) => (
  <TransitionHandler
    defaultStyles={{
      transition: `opacity ${transitionDurationMs}ms ${transitionTimingFunction}`,
      opacity: 0,
      position: 'fixed',
      zIndex: layers.blanket(),
    }}
    transitionStyles={{
      entering: { opacity: 0 },
      entered: { opacity: 1 },
    }}
    {...props}
  />
);

export const Slide = ({
  shouldUnmountOnExit = true,
  openFromRight,
  ...props
}: TransitionProps) => (
  <TransitionHandler
    openFromRight={openFromRight}
    defaultStyles={{
      transition:
        `transform ${transitionDurationMs}ms ${transitionTimingFunction}, ` +
        `width ${transitionDurationMs}ms ${transitionTimingFunction}`,
      transform: !openFromRight
        ? 'translate3d(-100%,0,0)'
        : 'translate3d(100%,0,0)',
      flexDirection: !openFromRight ? 'row' : 'row-reverse',
    }}
    transitionStyles={{
      // Unset transform so we do not create a new stacking context for fixed-position children - NAV-159
      entered: { transform: null },
      exited: {
        transform: !openFromRight
          ? 'translate3d(-100%,0,0)'
          : 'translate3d(100%,0,0)',
      },
    }}
    transitionProps={{
      ...defaultTransitionProps,
      ...{ unmountOnExit: shouldUnmountOnExit },
    }}
    {...props}
  />
);
