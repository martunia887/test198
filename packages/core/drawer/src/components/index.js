// @flow

import React, { Children, Component, Fragment, type Node } from 'react';
import { canUseDOM } from 'exenv';
import Portal from '@atlaskit/portal';
import { ThemeProvider } from 'styled-components';
import { TransitionGroup, Transition } from 'react-transition-group';
import {
  createAndFireEvent,
  withAnalyticsEvents,
  withAnalyticsContext,
  type WithAnalyticsEventsProps,
} from '@atlaskit/analytics-next';
import Blanket from '@atlaskit/blanket';
import {
  name as packageName,
  version as packageVersion,
} from '../version.json';
import drawerItemTheme from '../theme/drawer-item-theme';
import DrawerPrimitive from './primitives';
import { Fade } from './transitions';
import type { CloseTrigger, DrawerProps } from './types';

const OnlyChild = ({ children }) => Children.toArray(children)[0] || null;

const createAndFireEventOnAtlaskit = createAndFireEvent('atlaskit');

const createAndFireOnClick = (
  createAnalyticsEvent: $PropertyType<
    WithAnalyticsEventsProps,
    'createAnalyticsEvent',
  >,
  trigger: CloseTrigger,
) =>
  createAndFireEventOnAtlaskit({
    action: 'dismissed',
    actionSubject: 'drawer',
    attributes: {
      componentName: 'drawer',
      packageName,
      packageVersion,
      trigger,
    },
  })(createAnalyticsEvent);

export class DrawerBase extends Component<
  DrawerProps,
  { renderPortal: boolean },
> {
  static defaultProps = {
    width: 'narrow',
  };

  state = {
    renderPortal: false,
  };

  body = canUseDOM ? document.querySelector('body') : undefined;

  componentDidMount() {
    const { isOpen } = this.props;

    if (isOpen) {
      window.addEventListener('keydown', this.handleKeyDown);
    }
  }

  componentWillUnmount() {
    window.removeEventListener('keydown', this.handleKeyDown);
  }

  componentDidUpdate(prevProps: DrawerProps) {
    const { isOpen } = this.props;
    if (isOpen !== prevProps.isOpen) {
      if (isOpen) {
        window.addEventListener('keydown', this.handleKeyDown);
      } else {
        window.removeEventListener('keydown', this.handleKeyDown);
      }
    }
  }

  handleBlanketClick = (event: SyntheticMouseEvent<*>) => {
    this.handleClose(event, 'blanket');
  };

  handleBackButtonClick = (event: SyntheticMouseEvent<*>) => {
    this.handleClose(event, 'backButton');
  };

  handleClose = (
    event: SyntheticKeyboardEvent<*> | SyntheticMouseEvent<*>,
    trigger: CloseTrigger,
  ) => {
    const { createAnalyticsEvent, onClose } = this.props;

    const analyticsEvent = createAndFireOnClick(createAnalyticsEvent, trigger);

    if (onClose) {
      onClose(event, analyticsEvent);
    }
  };

  handleKeyDown = (event: SyntheticKeyboardEvent<*>) => {
    const { isOpen, onKeyDown } = this.props;

    if (event.key === 'Escape' && isOpen) {
      this.handleClose(event, 'escKey');
    }
    if (onKeyDown) {
      onKeyDown(event);
    }
  };

  render() {
    if (!this.body) {
      return null;
    }
    const {
      isOpen,
      children,
      icon,
      width,
      shouldUnmountOnExit,
      onCloseComplete,
    } = this.props;

    return (
      <Transition
        in={isOpen}
        timeout={{ enter: 0, exit: 220 }}
        mountOnEnter
        unmountOnExit
      >
        <Portal zIndex="unset">
          <TransitionGroup component={OnlyChild}>
            <Fragment>
              {/* $FlowFixMe the `in` prop is internal */}
              <Fade in={isOpen}>
                <Blanket isTinted onBlanketClicked={this.handleBlanketClick} />
              </Fade>
              <DrawerPrimitive
                icon={icon}
                in={isOpen}
                onClose={this.handleBackButtonClick}
                onCloseComplete={onCloseComplete}
                width={width}
                shouldUnmountOnExit={shouldUnmountOnExit}
              >
                {children}
              </DrawerPrimitive>
            </Fragment>
          </TransitionGroup>
        </Portal>
      </Transition>
    );
  }
}

export const DrawerItemTheme = (props: { children: Node }) => (
  <ThemeProvider theme={drawerItemTheme}>{props.children}</ThemeProvider>
);

export * from './skeletons';
export * from './item-group';

export default withAnalyticsContext({
  componentName: 'drawer',
  packageName,
  packageVersion,
})(withAnalyticsEvents()(DrawerBase));
