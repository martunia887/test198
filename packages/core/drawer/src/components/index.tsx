import React, {
  Children,
  Component,
  Fragment,
  FC,
  SyntheticEvent,
} from 'react';
import { canUseDOM } from 'exenv';
import Portal from '@atlaskit/portal';
import { ThemeProvider } from 'styled-components';
import { TransitionGroup } from 'react-transition-group';
import {
  createAndFireEvent,
  withAnalyticsEvents,
  withAnalyticsContext,
  CreateUIAnalyticsEventSignature,
} from '@atlaskit/analytics-next';
import Blanket from '@atlaskit/blanket';
import {
  name as packageName,
  version as packageVersion,
} from '../version.json';
import drawerItemTheme from '../theme/drawer-item-theme';
import DrawerPrimitive from './primitives';
import { Fade } from './transitions';
import { CloseTrigger, DrawerProps } from './types';

const OnlyChild: FC<any> = ({ children }) =>
  Children.toArray(children)[0] || null;

const createAndFireEventOnAtlaskit = createAndFireEvent('atlaskit');

const createAndFireOnClick = (
  createAnalyticsEvent: CreateUIAnalyticsEventSignature,
  trigger: CloseTrigger,
) =>
  createAndFireEventOnAtlaskit({
    action: 'dismissed',
    actionSubject: 'drawer',
    attributes: {
      componentName: 'drawer',
      packageName,
      packageVersion,
      trigger,
    },
  })(createAnalyticsEvent);

export class DrawerBase extends Component<DrawerProps> {
  static defaultProps = {
    width: 'narrow',
  };

  body = canUseDOM ? document.querySelector('body') : undefined;

  componentDidMount() {
    const { isOpen } = this.props;

    if (isOpen) {
      window.addEventListener('keydown', this
        .handleKeyDown as EventListenerOrEventListenerObject);
    }
  }

  componentWillUnmount() {
    window.removeEventListener('keydown', this.handleKeyDown);
  }

  componentDidUpdate(prevProps: DrawerProps) {
    const { isOpen } = this.props;
    if (isOpen !== prevProps.isOpen) {
      if (isOpen) {
        window.addEventListener('keydown', this.handleKeyDown);
      } else {
        window.removeEventListener('keydown', this.handleKeyDown);
      }
    }
  }

  private handleBlanketClick = (event: SyntheticEvent) => {
    this.handleClose(event, 'blanket');
  };

  private handleBackButtonClick = (event: SyntheticEvent) => {
    this.handleClose(event, 'backButton');
  };

  private handleClose = (event: SyntheticEvent, trigger: CloseTrigger) => {
    const { createAnalyticsEvent, onClose } = this.props;

    if (createAnalyticsEvent) {
      const analyticsEvent = createAndFireOnClick(
        createAnalyticsEvent,
        trigger,
      );

      if (onClose) {
        onClose(event, analyticsEvent);
      }
    }
  };

  handleKeyDown = (event: KeyboardEvent) => {
    const { isOpen, onKeyDown } = this.props;

    if (event.key === 'Escape' && isOpen) {
      this.handleClose((event as unknown) as React.KeyboardEvent, 'escKey');
    }
    if (onKeyDown) {
      onKeyDown((event as unknown) as React.KeyboardEvent);
    }
  };

  render() {
    if (!this.body) {
      return null;
    }
    const {
      isOpen,
      children,
      icon,
      width,
      shouldUnmountOnExit,
      onCloseComplete,
    } = this.props;
    return (
      <Portal zIndex="unset">
        <TransitionGroup component={OnlyChild}>
          <Fragment>
            {/* $FlowFixMe the `in` prop is internal */}
            <Fade in={isOpen}>
              <Blanket isTinted onBlanketClicked={this.handleBlanketClick} />
            </Fade>
            <DrawerPrimitive
              icon={icon}
              in={isOpen}
              onClose={this.handleBackButtonClick}
              onCloseComplete={onCloseComplete}
              width={width}
              shouldUnmountOnExit={shouldUnmountOnExit}
            >
              {children}
            </DrawerPrimitive>
          </Fragment>
        </TransitionGroup>
      </Portal>
    );
  }
}

export const DrawerItemTheme: FC = props => (
  <ThemeProvider theme={drawerItemTheme}>{props.children}</ThemeProvider>
);

export * from './skeletons';
export * from './item-group';

export default withAnalyticsContext({
  componentName: 'drawer',
  packageName,
  packageVersion,
})(withAnalyticsEvents()(DrawerBase));
