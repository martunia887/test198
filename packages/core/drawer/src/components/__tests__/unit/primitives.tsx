import React, { FC } from 'react';
import { mount, shallow } from 'enzyme';
import ArrowLeft from '@atlaskit/icon/glyph/arrow-left';

import DrawerPrimitive from '../../primitives';
import { Slide } from '../../transitions';
import { DrawerWidth, IconProps } from 'src/components/types';

const DrawerContent: FC<any> = () => <code>Drawer contents</code>;

describe('Drawer primitive', () => {
  const commonProps = {
    width: 'wide' as DrawerWidth,
    in: true,
    shouldUnmountOnExit: false,
  };

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should render given icon in large size if exists', () => {
    const props = { ...commonProps, icon: () => <span>Icon</span> };
    const wrapper = mount(
      <DrawerPrimitive {...props}>
        <DrawerContent />
      </DrawerPrimitive>,
    );
    expect((wrapper.find(props.icon).props() as IconProps).size).toBe('large');
  });

  it('should render arrow left if icon prop does NOT exist', () => {
    const props = { ...commonProps };
    const wrapper = mount(
      <DrawerPrimitive {...props}>
        <DrawerContent />
      </DrawerPrimitive>,
    );

    expect(wrapper.find(ArrowLeft).length).toBe(1);
  });

  it('should remount the node if receives shouldUnmountOnExit prop', () => {
    const props = { ...commonProps, shouldUnmountOnExit: true };
    const wrapper = mount(
      <DrawerPrimitive {...props}>
        <DrawerContent />
      </DrawerPrimitive>,
    );
    expect(wrapper.getDOMNode()).not.toBe(null);

    wrapper.setProps({ in: false });
    jest.runTimersToTime(20000);
    wrapper.update();

    expect(wrapper.getDOMNode()).toBe(null);
  });

  it('should NOT remount the node if shouldUnmountOnExit is false', () => {
    const props = { ...commonProps };
    const wrapper = mount(
      <DrawerPrimitive {...props}>
        <DrawerContent />
      </DrawerPrimitive>,
    );
    expect(wrapper.getDOMNode()).not.toBe(null);

    wrapper.setProps({ in: false });
    jest.runTimersToTime(20000);
    wrapper.update();

    expect(wrapper.getDOMNode()).not.toBe(null);
  });

  it('should render with medium width', () => {
    const props = { ...commonProps, width: 'medium' as DrawerWidth };
    const wrapper = mount(
      <DrawerPrimitive {...props}>
        <DrawerContent />
      </DrawerPrimitive>,
    );
    expect(wrapper.find(DrawerPrimitive).props().width).toBe('medium');
  });

  it('should call onClose when the icon is clicked', () => {
    const onClose = jest.fn();
    const props = { ...commonProps, onClose };
    const wrapper = shallow(
      <DrawerPrimitive {...props}>
        <DrawerContent />
      </DrawerPrimitive>,
    );

    const event = { target: 'button' };
    const callsBeforeIconClick: any[] = [].concat(onClose.mock.calls as any);
    (wrapper.find('IconWrapper').prop('onClick') as (args: any) => void)(event);
    const callsAfterIconClick = onClose.mock.calls;

    expect({ callsBeforeIconClick, callsAfterIconClick }).toEqual({
      callsBeforeIconClick: [],
      callsAfterIconClick: [[event]],
    });
  });

  it('should call onCloseComplete when the Slide has exited', () => {
    const onCloseComplete = jest.fn();
    const props = { ...commonProps, in: true, onCloseComplete };
    const wrapper = shallow(
      <DrawerPrimitive {...props}>
        <DrawerContent />
      </DrawerPrimitive>,
    );

    const node = 'div';
    const callsBeforeExited = [].concat(onCloseComplete.mock.calls as any);
    wrapper
      .find(Slide)
      .props()
      .onExited(node);
    const callsAfterExited = onCloseComplete.mock.calls;

    expect({
      callsBeforeExited,
      callsAfterExited,
    }).toEqual({
      callsBeforeExited: [],
      callsAfterExited: [[node]],
    });
  });
});
