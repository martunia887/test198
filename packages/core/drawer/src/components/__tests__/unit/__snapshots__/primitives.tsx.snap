// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer primitive should allow content wrapper to be themed 1`] = `
.emotion-0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-top: 0;
  overflow: auto;
}

<div
  className="emotion-0"
/>
`;
